core/identity-spi/src/main/java/software/amazon/awssdk/identity/spi/ResolveIdentityRequest.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.identity.spi;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.builder.SdkBuilder;
+
+/**
+ * A request to resolve an {@link Identity}.
+ *
+ * The Identity may be determined for each request based on properties of the request (e.g. different credentials per bucket
+ * for S3).
+ *
+ * @see IdentityProvider
+ */
+@SdkProtectedApi
+@Immutable
+@ThreadSafe
+public final class ResolveIdentityRequest {
+
+    private final Map<IdentityProperty<?>, ?> properties;
+
+    private ResolveIdentityRequest(Map<IdentityProperty<?>, Object> properties) {
+        this.properties = new HashMap<>(properties);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public <T> T property(IdentityProperty<T> property) {
+        return (T) properties.get(property);
+    }
+
+    @Override
+    public String toString() {
+        return ToString.builder("ResolveIdentityRequest")
+                       .add("properties", properties)
+                       .build();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ResolveIdentityRequest that = (ResolveIdentityRequest) o;
+        return Objects.equals(properties, that.properties);
+    }
+
+    @Override
+    public int hashCode() {
+        return 31 + Objects.hashCode(properties);
```

Comment:
I think this should be `32 * Objects.hashCode(properties);` instead.

Reply:
Sorry, `31 * ...`, i.e., not `+`.

Reply:
Actually, I think since there is only 1 field in this class, I can just `return Objects.hashCode(properties);`. Will update in new PR as this is merged.

Reply:
Updating in https://github.com/aws/aws-sdk-java-v2/pull/3814